[Doc Improvement][Navigation within tab app] (#7381)
MicrosoftDocs/msteams-docs: msteams-platform/tabs/how-to/personal-tab-navigation.md

---
title: Personal tab app navigation
description: Learn about the options for navigation within a personal tab app using the TeamsJS library.
ms.date: 10/26/2022
ms.topic: conceptual
ms.custom: m365apps
ms.localizationpriority: medium
---

# Navigate within a personal tab app

The TeamsJS library provides support for navigation within a personal app. This article discusses the options available, which includes types of navigation such as that between tabs within the app or through use of Teams UI components like the back button.

> [!INCLUDE [sdk-include](~/includes/sdk-include.md)]

Reviewer: Would you mind updating the include to **2.x.x** as part of this PR? Thanks!

Author: Sure thing. Updated the include to say 2.x.x. Will be in next commit shortly

> Personal tabs provide a great way to enhance the Microsoft Teams experience. You can provide the user access to your web application right within Teams using personal tabs, without having to sign in again. For more information about personal tabs and how you can extend them across Microsoft 365 products, see [Build tabs for Teams](~/tabs/what-are-tabs.md) and [Extend a Teams personal tab across Microsoft 365](~/m365-apps/extend-m365-teams-personal-tab.md).

Reviewer 2: ```suggestion
Personal tabs provide a great way to enhance Microsoft Teams experience. You can provide user access to your web application right within Teams using personal tabs, without having to sign in. For more information about personal tabs and how you can extend them across Microsoft 365 products, see [Build tabs for Teams](~/tabs/what-are-tabs.md) and [Extend a Teams personal tab across Microsoft 365](~/m365-apps/extend-m365-teams-personal-tab.md).
```

> Personal tabs provide a great way to enhance the Microsoft Teams experience. Using personal tabs, you can provide a user access to their application right within Teams, without the user having to leave the experience or sign in again. For more information about personal tabs and how you can extend them across Microsoft 365 products, see [Build tabs for Teams](~/tabs/what-are-tabs.md) and [Extend a Teams personal tab across Microsoft 365](~/m365-apps/extend-m365-teams-personal-tab.md).

Reviewer: This reads a little ambiguous to me (seems more written towards admins rather than developers): *Using personal tabs, you can provide a user access to their application right within Teams* 
. . . how about: 
With personal tabs, you can provide users access to **your web** application right within Teams

Reviewer 3: This has been updated

## Navigate between tabs

The [pages](/javascript/api/@microsoft/teams-js/pages?view=msteams-client-js-latest&preserve-view=true) capability of the TeamsJS library provides support for navigation between tabs within a personal app. Specifically, the [pages.currentApp](/javascript/api/@microsoft/teams-js/pages.currentapp?view=msteams-client-js-latest&preserve-view=true) namespace offers a function `navigateTo(NavigateWithinAppParams)` to allow navigation to a specific tab within the current app and a function `navigateToDefaultPage()` to navigate to the first tab defined in the app's manifest.

The following code illustrates how to navigate to a specific page:

```js

> if (pages.currentApp.isSupported()) {

Reviewer 4: So I was checking and saw that teams Hub has yet to support the pages.currentApp namespace even though that is released in the latest version of the TeamsJS SDK. I talked to an engineering manager who said that it would be best to make two sections(one for hubs that do support and one for those who do not). This is because Outlook and Office are supposed to have support for this very soon.

Reviewer 3: @vikramtha
Do you require them to be appearing in 2 different tabs? Hubs that do support and that do not.

Author: @vikramtha @v-amprasad - I think this might be worth a conversation on later today  at our sync @vikramtha. Given we show doing the isSupported check for the varying support in the different products, I would argue against breaking it out. Setting a practice of breaking out support for each teamsjs call is just going to be a mess and hard to keep track of. I'd rather see it as part of the API ref or as a single table somewhere if anything. If not having the Teams support yet is an issue, I'd suggest holding off on the article or potentially publishing this info in a different way (part of our Tab extension articlie) until the full support is there.

    const navPromise = pages.currentApp.navigateTo({pageId: <pageId>, subPageId:<subPageId>});
    navPromise.
        then((result) => {/*Successful navigation*/}).
        catch((error) => {/*Failed navigation*/});
}
else {/*Handle situation where capability isn't supported*/}
```

The following code illustrates how to navigate to the app's default tab:

```js
if (pages.currentApp.isSupported()) {
    const navPromise = pages.currentApp.navigateToDefaultPage();
    navPromise.
        then((result) => {/*Successful navigation*/}).
        catch((error) => {/*Failed navigation*/});
}
else {/*Handle situation where capability isn't supported*/}
```

> It is also possible to use the `pages.navigateToApp()` function, `pages.tabs.navigateToTab()` function, or deep links for tab navigation, as shown in [Create deep links](~/concepts/build-and-test/deep-links.md#navigate-within-your-app), however the preferred method for apps that are extended across Microsoft 365 is to use the APIs in the `pages.currentApp` namespace.

Reviewer: Would be good to avoid passive voice (per style guide). How about:

**You can also** use the `pages.navigateToApp()` function, `pages.tabs.navigateToTab()` function, or deep links for tab navigation,

Reviewer 3: This has been changed as per style guide

> It is also possible to use the `pages.navigateToApp()` function, `pages.tabs.navigateToTab()` function, or deep links for tab navigation, as shown in [Create deep links](~/concepts/build-and-test/deep-links.md#navigate-within-your-app), however the preferred method for apps that are extended across Microsoft 365 is to use the APIs in the `pages.currentApp` namespace.

Reviewer: Best practice for all Teams apps (not just M365 extended) using TeamsJS 2.x.x is to use the new navigation APIs. Let's generalize this recommendation.

Reviewer 3: this has been updated.

Author: Given the new APIs are still marked Beta, I've removed this sentence. Working with team to determine best practice/recommendation and will update once I've determined.

## Configure back button navigation

> When a personal app has multiple tabs, a user can use the Microsoft 365 host app's back button to go backwards through the navigational history. However that history doesn't include actions a user performs within a tab. If you need to enhance the back button experience, you can maintain your own internal navigation stack and configure a custom handler for back button selections. This is accomplished through the `registerBackButtonHandler()` function in the [`pages.backStack`](/javascript/api/@microsoft/teams-js/pages.backstack?view=msteams-client-js-latest&preserve-view=true) namespace.

Reviewer 2: ```suggestion
When a personal app has multiple tabs, a user can use the Microsoft 365 host app's back button to go backwards through the navigational history. However, the history doesn't include the actions a user performs within a tab. If you want to enhance the back button experience, you can maintain your own internal navigation stack and configure a custom handler for back button selections. This can be accomplished through the `registerBackButtonHandler()` function in the [`pages.backStack`](/javascript/api/@microsoft/teams-js/pages.backstack?view=msteams-client-js-latest&preserve-view=true) namespace.
```

> When a personal app has multiple tabs, a user can use the Microsoft 365 host app's back button to go backwards through the navigational history. However, that history does not include actions a user might have taken within a tab. If you would like to better control the back button experience, it is possible to maintain your own internal navigation stack and configure a custom handler for back button presses. This is accomplished through the `registerBackButtonHandler()` function in the [pages.backStack](/javascript/api/@microsoft/teams-js/pages.backstack?view=msteams-client-js-latest&preserve-view=true) namespace.

Reviewer: Instead of *it is possible to* (passive voice), how about just **you can**

> After you register the handler, it helps you to address the navigational request before the system takes action. If the handler is able to manage the request, it returns `true` so that the system knows no further action is necessary. If the internal stack is empty, it returns `false` so that the system can call the `navigateBack()` function instead and take the appropriate action.

Reviewer 4: The handler doesn't **return** any values or booleans. It is either set to true or false

Author: @vikramtha - Am I misunderstanding the [handler syntax](https://learn.microsoft.com/en-us/javascript/api/@microsoft/teams-js/pages.backstack?view=msteams-client-js-latest#@microsoft-teams-js-pages-backstack-registerbackbuttonhandler)? It looks like we expect it to return a boolean.

Reviewer 4: Ah my bad you are correct. The function registerBackButtonHandler doesn't **return** anything but the handler itself returns the boolean. I kept reading "handler" as the entire function not the parameter

> After you register the handler, it helps you to address the navigational request before the system takes action. If the handler is able to manage the request, it is set to `true` so that the system knows no further action is necessary. If the internal stack is empty, it is set to `false` so that the system can call the `navigateBack()` function instead and take the appropriate action.

Reviewer 2: ```suggestion
After you register the handler, it helps you to address the navigational request before the system takes action. If the handler is able to manage the request, it's set to `true` so that the system knows no further action is necessary. If the internal stack is empty, it's set to `false` so that the system can call the `navigateBack()` function instead and take the appropriate action.
```

## Return focus to host app

> After the user starts using elements within a tab, by default focus remains with the elements of your iframe until the user selects outside of it. If it's a part of the user navigating with keyboard shortcuts (tab or F6), you can again focus to the host app. You can focus on the host app by using the [`pages.returnFocus()`](/javascript/api/@microsoft/teams-js/pages?view=msteams-client-js-latest#@microsoft-teams-js-pages-returnfocus&preserve-view=true) function. The `returnFocus()` function accepts a boolean indicating the direction to advance focus within the host, `true` for forward and `false` for backwards. Generally, forward results in the search bar being highlighted and backwards results in the app bar being highlighted.

Reviewer 2: ```suggestion
After the user starts using elements within a tab, by default, the focus remains with the elements of your iFrame until the user selects outside of it. If the iFrame is a part of the user navigating with keyboard shortcuts (the Tab key or the F6 key), you can focus on the host app. You can focus on the host app by using the [`pages.returnFocus()`](/javascript/api/@microsoft/teams-js/pages?view=msteams-client-js-latest#@microsoft-teams-js-pages-returnfocus&preserve-view=true) function. The `returnFocus()` function accepts a Boolean indicating the direction to advance focus within the host app; `true` for forward and `false` for backwards. Generally, forward highlights the search bar and backwards highlights the app bar.
```

> Once a user starts using elements within a tab, by default focus remains with the elements of your iFrame until the user clicks outside of it. If, as part of the user navigating with keyboard shortcuts (tab or F6) you'd like to return focus to the host app, you can use the [pages.returnFocus()](/javascript/api/@microsoft/teams-js/pages?view=msteams-client-js-latest#@microsoft-teams-js-pages-returnfocus&preserve-view=true) function. The `returnFocus()` function accepts a boolean indicating the direction to advance focus within the host, `true` for forward and `false` for backwards. Generally, forward results in the search bar being highlighted and backwards results in the app bar being highlighted.

Reviewer: I see it's camelCased in another Teams doc, but iframe should really be lowercase

Author: Thanks, updated.

> ## See also

Reviewer: Wish we had a sample to add here, but I don't think any of the current ones demonstrate the new pages.currentApp APIs. As a separate project, might be worth investigating what it'd take to include them in the canonical [TeamsJS sample app](ttps://github.com/OfficeDev/microsoft-teams-library-js/tree/main/apps/sample-app), which is intended to demo navigation (amonth other things)
